modin/core/execution/unidist/implementations/pandas_on_unidist/io/io.py

```diff
@@ -0,0 +1,323 @@
+        kwargs["if_exists"] = "append"
+        columns = qc.columns
+
+        def func(df):
+            """
+            Override column names in the wrapped dataframe and convert it to SQL.
+
+            Notes
+            -----
+            This function returns an empty ``pandas.DataFrame`` because ``apply_full_axis``
+            expects a Frame object as a result of operation (and ``to_sql`` has no dataframe result).
+            """
+            df.columns = columns
+            df.to_sql(**kwargs)
+            return pandas.DataFrame()
+
+        # Ensure that the metadata is syncrhonized
+        qc._modin_frame._propagate_index_objs(axis=None)
+        result = qc._modin_frame.apply_full_axis(1, func, new_index=[], new_columns=[])
+        # FIXME: we should be waiting for completion less expensievely, maybe use _modin_frame.materialize()?
+        result.to_pandas()  # blocking operation
+
+    @staticmethod
+    def _to_csv_check_support(kwargs):
+        """
+        Check if parallel version of ``to_csv`` could be used.
+
+        Parameters
+        ----------
+        kwargs : dict
+            Keyword arguments passed to ``.to_csv()``.
+
+        Returns
+        -------
+        bool
+            Whether parallel version of ``to_csv`` is applicable.
+        """
+        path_or_buf = kwargs["path_or_buf"]
+        compression = kwargs["compression"]
+        if not isinstance(path_or_buf, str):
+            return False
+        # case when the pointer is placed at the beginning of the file.
+        if "r" in kwargs["mode"] and "+" in kwargs["mode"]:
+            return False
+        # encodings with BOM don't support;
+        # instead of one mark in result bytes we will have them by the number of partitions
+        # so we should fallback in pandas for `utf-16`, `utf-32` with all aliases, in instance
+        # (`utf_32_be`, `utf_16_le` and so on)
+        if kwargs["encoding"] is not None:
+            encoding = kwargs["encoding"].lower()
+            if "u" in encoding or "utf" in encoding:
+                if "16" in encoding or "32" in encoding:
+                    return False
+        if compression is None or not compression == "infer":
+            return False
+        if any((path_or_buf.endswith(ext) for ext in [".gz", ".bz2", ".zip", ".xz"])):
+            return False
+        return True
+
+    @classmethod
+    def to_csv(cls, qc, **kwargs):
+        """
+        Write records stored in the `qc` to a CSV file.
+
+        Parameters
+        ----------
+        qc : BaseQueryCompiler
+            The query compiler of the Modin dataframe that we want to run ``to_csv`` on.
+        **kwargs : dict
+            Parameters for ``pandas.to_csv(**kwargs)``.
+        """
+        if not cls._to_csv_check_support(kwargs):
+            return UnidistIO.to_csv(qc, **kwargs)
+
+        signals = SignalActor.remote(len(qc._modin_frame._partitions) + 1)
+
+        def func(df, **kw):
```

Comment:
IIUC, you want to combine some code for unidist and Ray engines. If this is what you mean, this is supposed to be a separate issue. As the final goal, it would be perfect for us to move from `PandasOnRay`/`Dask`/`Etc`. to `PandasOnUnidist`.

Reply:
I see - that makes a lot of sense.